### PR TITLE
Introduces structs

### LINKED ISSUE #103: Add cache support to Graph module
### 🎛 Description

**Choose one:** 🙋 feature request

Update Graph module to use cache. At the moment it will re-issue requests for the same information.

### LINKED ISSUE #104: Extract Graph#get to Service module.
### 🎛 Description

**Choose one:** 🙋 feature request?

Let's pull Graph#get into the service module to reduce duplication.

### LINKED ISSUE #106: Cannot check properties Azure resource properties from azurerm_virtual_machine
### 🎛 Description

🐛 bug report

I am expecting this control to work:
```
control 'azure_virtual_machine' do
  impact 0.7
  title 'Azure VM check'
  describe azurerm_virtual_machine(resource_group: 'my_rg', name: 'host') do
    its('properties.storageProfile.imageReference.publisher') { should eq "OpenLogic" }
  end
end
```
but I actually get:
```
  ×  test: 'host' Virtual Machine
     ×  'host' Virtual Machine properties.storageProfile.imageReference.publisher
     undefined method `storageProfile' for #<Hash:0x000055d22eaf45c8>
```

### 🌍 InSpec and Platform Version

Inspec Version: 2.2.50
ruby 2.5.1p57
Tried both locally on my Mac OS 10.13.6 and on the official inspec docker image Alpine Linux v3.7

### 🤔 Replication Case

- Setup Environment Azure credentials 
- Create an inspec profile
- use the following control:
```
control 'azure_virtual_machine' do
  impact 0.7
  title 'Azure VM check'
  describe azurerm_virtual_machine(resource_group: 'my_rg', name: 'host') do
    its('properties.storageProfile.imageReference.publisher') { should eq "OpenLogic" }
  end
end
```
in inspec.yml:
```
name: azure-create-virtualmachine
title: InSpec Profile
maintainer: The Authors
copyright: The Authors
copyright_email: you@example.com
license: Apache-2.0
summary: An InSpec Compliance Profile
version: 0.1.0
inspec_version: '>= 2.2.7'
depends:
  - name: inspec-azure
    git: https://github.com/inspec/inspec-azure.git
supports:
  - platform: azure
```

run `inspec exec . -t azure://`

### 💁 Possible Solutions

N/A

### 💻 Stacktrace

N/A

### PR DESCRIPTION
**Pull request by dmccown**
### Description

This change introduces deep structs in resources. With this you may now access deeper properties in the resource. E.g. `properties.property1.property2` may be accessed assuming all those keys exist for the resource. If a response excludes a key it may not exist.

### Issues Resolved

fixes #106 
fixes #103 
fixes #104 

### Check List

- [x] New functionality includes tests
- [x] All tests pass
- [x] `rake lint` passes
- [ ] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

Co-authored-by: Trevor Bramble tbramble@chef.io
Co-authored-by: David McCown dmccown@chef.io

### CODE REVIEW

File: libraries/azurerm_monitor_activity_log_alert.rb [hunk 1]
```diff
@@ -50,6 +44,6 @@ def to_s
   # @param [Hash] 'allOf' conditions from response properties
   # @return [Array] of operation strings
   def collect_operations(conditions)
-    conditions.find_all { |x| x['field'] == 'operationName' }.collect { |x| x['equals'] }
+    conditions.find_all { |x| x.field == 'operationName' }.collect(&:equals)
```
Comment (TrevorBramble) [inline libraries/azurerm_monitor_activity_log_alert.rb]: Symbol-to-proc validates this entire change.

File: libraries/support/azure/service.rb [hunk 2]
```diff
@@ -28,6 +28,20 @@ def for_subscription(subscription_id, override: false)
       set_reader(:subscription_id, subscription_id, override)
     end
 
+    def structify(data)
```
Comment (TrevorBramble) [inline libraries/support/azure/service.rb]: It only occurs to me with the distance of time that maybe I should've added a little documenting comment to this method. What do you think?

File: libraries/support/azure/service.rb [hunk 3]
```diff
@@ -52,12 +66,16 @@ def set_reader(name, value, override)
...
 
-    def get(url:, api_version:)
+    def get(url:, api_version:, error: nil, unwrap: nil)
       confirm_configured!
 
       cache.fetch(url) do
         body = rest_client.get(url, params: { 'api-version' => api_version }).body
-        body.fetch('value', body)
+
+        error&.call(body)
```
Comment (TrevorBramble) [inline libraries/support/azure/service.rb]: You know, I've disregarded the shorthand for `call` but `error&.(body)` seems much more to the point than `error&.call(body)`. We don't have to change anything here, but food for thought.

File: libraries/support/azure/service.rb [hunk 4]
```diff
@@ -52,12 +66,16 @@ def set_reader(name, value, override)
...
+
+        return unwrap.call(body, api_version) unless unwrap.nil?
+        structify(body.fetch('value', body))
```
Comment (TrevorBramble) [inline libraries/support/azure/service.rb]: There's a sort of implicit `if/else` happening here that I think might read better if it were explicit.

```ruby
def get(url:, api_version:, error: nil, unwrap: nil)
  # ...

  if unwrap.respond_to?(:call)
    unwrap.call(body, api_version)
  else
    structify(body.fetch('value', body))
  end
end

File: libraries/support/azure/service.rb [hunk 5]
```diff
@@ -52,12 +66,16 @@ def set_reader(name, value, override)
       self
     end
 
-    def get(url:, api_version:)
+    def get(url:, api_version:, error: nil, unwrap: nil)
```
Comment (TrevorBramble) [inline libraries/support/azure/service.rb]: I wonder about this name, `error`. It's really an optional error handler, right?

File: test/unit/support/azure/service_test.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
+require_relative '../../test_helper'
+require_relative '../../../../libraries/support/azure/service'
+
+describe TestCredentials do
+  before do
+    @service = Object.new
+    @service.extend(Azure::Service)
```
Comment (TrevorBramble) [inline test/unit/support/azure/service_test.rb]: One line should suffice.
```ruby
@service = Object.new.extend(Azure::Service)
```

File: test/unit/support/azure/service_test.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  it 'converts simple hashes to structs' do
+    struct = @service.structify({ name: 'John', age: 20 })
```
Comment (TrevorBramble) [inline test/unit/support/azure/service_test.rb]: Do we actually need the curly braces? (Doesn't Rubocop ding you for using them when they could be omitted?)

File: test/unit/support/azure/service_test.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    assert_equal('John', struct.name)
+    assert_equal(20, struct.age)
+  end
+
+  it 'converts arrays to structs' do
+    struct = @service.structify([{ name: 'John', age: 20 },
+                                 { name: 'Jane', age: 25 }])
```
Comment (TrevorBramble) [inline test/unit/support/azure/service_test.rb]: Not going to fight about it, but I find extracting the setup to `let` statements with explaining variables helps with overall readability both by coercing you to give the different data distinguishing names, but also by removing their setup from the comparisons.

File: libraries/azurerm_resource.rb [hunk 9]
```diff
@@ -42,9 +42,39 @@ def credentials
...
-class AzurermPluralResource < AzurermResource; end
+class AzurermPluralResource < AzurermResource
+  def add_key(struct, key, value)
+    Struct.new(*struct.members << key) do
+      def key?(key)
+        members.include?(key)
+      end
+
+      alias_method :keys, :members
+    end.new(*struct.values << value)
```
[the remark this answers is not in the document]
Comment (dmccown) [reply libraries/azurerm_resource.rb]: 👍 on pulling out it's own module. The test I wrote only tests structify and I had to make it public. It should be it's own module.

[the remark this answers is not in the document]
Comment (dmccown) [reply libraries/azurerm_resource.rb]: Sorry misunderstood what your preference was. That sounds reasonable as well.

**Pull request merged**